### PR TITLE
KOGITO-2207 Review AbstractProcessInstance to use WorkflowProcessInst…

### PR DESCRIPTION
**Pull request by cristianonicolai**
…ance interface avoiding multiples casts

Small refactor with a series of improvements to avoid multiple casts in AbstractProcessInstance. As it depends on WorkflowProcessInstance anyway. Moved some method to interface to also avoid dpendency on Impl class.

### CODE REVIEW

File: jbpm/jbpm-flow/src/main/java/org/jbpm/process/instance/ProcessInstance.java [hunk 1]
```diff
@@ -44,14 +44,14 @@
...
-
+    
     void setKnowledgeRuntime(InternalKnowledgeRuntime kruntime);
 
     InternalKnowledgeRuntime getKnowledgeRuntime();
 
     void start();
 
-    void start(String tigger);
+    void start(String trigger);
```
Comment (ruromero) [inline jbpm/jbpm-flow/src/main/java/org/jbpm/process/instance/ProcessInstance.java:54]: I prefer `tiger` :rofl:

File: jbpm/jbpm-flow/src/main/java/org/jbpm/workflow/instance/NodeInstanceContainer.java [hunk 2]
```diff
@@ -35,21 +37,39 @@
...
-    
+
     int getCurrentLevel();
-    
+
+    void setCurrentLevel(int level);
+
     NodeInstance getNodeInstance(String nodeInstanceId, boolean recursive);
 
+    default NodeInstance getNodeInstanceByNodeDefinitionId(final String nodeDefinitionId, NodeContainer nodeContainer) {
```
Comment (ruromero) [inline jbpm/jbpm-flow/src/main/java/org/jbpm/workflow/instance/NodeInstanceContainer.java:57]: why not just `getByDefinitionId` as it is the `NodeInstanceContainer` it can be assumed that the `get` will return a `NodeInstance`
The same could apply to the method above and call it just `get`

File: jbpm/jbpm-flow/src/main/java/org/jbpm/workflow/instance/impl/WorkflowProcessInstanceImpl.java [hunk 3]
```diff
@@ -380,6 +359,10 @@ public void setState(final int state, String outcome, Object faultData) {
     @Override
     public void setState(final int state, String outcome) {
         // TODO move most of this to ProcessInstanceImpl
+        if (state != ProcessInstance.STATE_ERROR) {
```
Comment (ruromero) [inline jbpm/jbpm-flow/src/main/java/org/jbpm/workflow/instance/impl/WorkflowProcessInstanceImpl.java:362]: Why is this needed? Can you add a unit test to support it?

File: jbpm/jbpm-flow/src/main/java/org/jbpm/workflow/instance/NodeInstanceContainer.java (hunk 2 above)
Comment (cristianonicolai) [reply jbpm/jbpm-flow/src/main/java/org/jbpm/workflow/instance/NodeInstanceContainer.java]: @ruromero renamed the new method to `getByDefinitionId` as suggested, but keeping old methods as is to avoid more unrelated changes.

File: jbpm/jbpm-flow/src/main/java/org/jbpm/workflow/instance/impl/WorkflowProcessInstanceImpl.java (hunk 3 above)
Comment (cristianonicolai) [reply jbpm/jbpm-flow/src/main/java/org/jbpm/workflow/instance/impl/WorkflowProcessInstanceImpl.java]: rollback change to not cause any behaviour change.

File: jbpm/jbpm-flow/src/main/java/org/jbpm/process/instance/ProcessInstance.java (hunk 1 above)
Comment (mswiderski) [reply jbpm/jbpm-flow/src/main/java/org/jbpm/process/instance/ProcessInstance.java:54]: same here tigger (double g to make it stronger) makes a lot of sense :)
